[torchbind] Add inductor support

angelayi opened the pull request:
Example inductor generated python code: [P1245776497](https://www.internalfb.com/phabricator/paste/view/P1245776497)
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123709



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

angelayi committed:
[torchbind] Add inductor support

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1210320502




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1210320502




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1210320502




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -913,7 +913,7 @@ class CompiledFxGraph:
     mutated_inputs: Set[str]
     mutated_input_idxs: Set[int]
-    constants: Dict[str, torch.Tensor]
+    constants: Dict[str, Union[torch.Tensor, torch._C.ScriptObject]]
```
Should we just add a separate, torchbind objects field here?

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -966,13 +971,20 @@ def add_expr_input(name, val):
             )
 
+            num_torchbind = 0
             for name, value in V.graph.constants.items():
```
can we just add torchbind_constants field instead ?

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -118,7 +118,13 @@ def set_env(arg):
         flattened_inputs = pytree.arg_tree_leaves(*args, **kwargs)
 
-        if self.unknown_value in flattened_inputs:
+        # We need to do this weird thing because in cases where flattened_inputs
+        # contains a ScriptObject, equality checking results in a type error if
+        # the types are different.
```
Is this something that can be fixed on ScriptObject ?

angelayi replied:
I'm not sure how? Because if object the equality check was type-erroring on had type "object". I'm not even sure what kind of object that was...

eellison replied:
Could we return false in the case that we would throw ? maybe file an issue for this ?

eellison commented on torch/_inductor/graph.py:
```diff
@@ -605,7 +606,9 @@ def get_buffer(self, buffer_name: str):
     def get_dtype(self, buffer_name: str):
         if buffer_name in self.constants:
-            return self.constants[buffer_name].dtype
+            buffer = self.constants[buffer_name]
```
+1 separate field for torchbind constants

eellison commented on torch/_inductor/graph.py:
```diff
@@ -876,7 +887,12 @@ def get_attr(self, target, args, kwargs):
-        if (
+        elif isinstance(value, torch._C.ScriptObject):
+            self.constants[target] = value
+            # self.constant_reprs[target] = ??
```
what is this ??

angelayi replied:
I'm not sure? I saw it when we're [creating tensor constants](https://fburl.com/code/sww1tsmp) and was wondering if this is needed here too

angelayi replied:
Oh I realized, it's used as a comment for when we declare the constants globally ([code pointer](https://www.internalfb.com/code/fbsource/[c1a73c1091ecd3250d74b396275058ea412f062c]/fbcode/caffe2/torch/_inductor/codegen/wrapper.py?lines=464-466%2C486-487)):
```
_tensor_constant0 = None  # constant_reprs["_tensor_constant0"] 
```

eellison commented on torch/_inductor/ir.py:
```diff
@@ -5181,13 +5185,14 @@ def __repr__(self):
         if isinstance(example_output, torch.Tensor):
             return example_output.device
         if isinstance(example_output, (list, tuple)):
-            devices = {FallbackKernel.find_device(None, x) for x in example_output}
+            device_set = {FallbackKernel.find_device(None, x) for x in example_output}
```
why this change ? sets are non deterministic so we should avoid

angelayi replied:
This is just a rename from devices -> device_set

eellison commented on torch/_inductor/ir.py:
```diff
@@ -6240,12 +6245,7 @@ def create(
-            kernel_layout,
-            _,
-        ) = _prepare_convolution_fusion_create(
+        (inputs, constant_args, kernel_layout, _,) = _prepare_convolution_fusion_create(
```
nit: formatting change

angelayi replied:
I think lintrunner complained if I didn't have this change 😅

eellison commented on torch/_inductor/ir.py:
```diff
@@ -7392,7 +7392,7 @@ def __init__(
     ):
         super().__init__(
-            NoneLayout(layout.device),
+            layout,
```
why did we need this change ?

angelayi replied:
I found that if layout is None, then during codegen the program doesn't flush correctly [here](https://fburl.com/code/0vuhx3hp), and then the generated python code will be incorrect.

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -951,6 +951,11 @@ def add_fake_input(name, shape, stride, device, dtype):
+        def add_torchbind_input(name, value):
+            import pickle
+
+            output.writeline(f"{name} = pickle.loads({pickle.dumps(value)!r})")
```
are pickle objects round trip serializable via pickle dumps? I thought with torchbind we did serialization via get state and set state.. should we use that ?

angelayi replied:
Yeah I thought the get_state and set_state are used to override the pickle logic?

eellison reviewed (commented):
A few comments. I think we'll need to add support for fx graph caching or skip similar to https://github.com/pytorch/pytorch/pull/123325#issue-comment-box

eellison reviewed (commented):
A few comments. I think we'll need to add support for fx graph caching or skip similar to https://github.com/pytorch/pytorch/pull/123325#issue-comment-box

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1210320502




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1210320502




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1234027981 (without benchmark_compiled_module) P1234029088 (with benchmark_compiled_module)

I also added a `load_constants` function in the wrapper module which gets called at the toplevel (suggested by shunting314). The purpose of this function is to load the constants (torchbind objects, constant tensors). Previously, this code was only in the benchmark_compiled_module function, to prevent the constants from being loaded on every `call`. However, that function is not called in the jit inductor path, causing the torchbind constants to not be loaded. So instead, I added the `load_constants` function which gets called in the toplevel module once.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1234027981 (without benchmark_compiled_module) P1234029088 (with benchmark_compiled_module)

I also added a `load_constants` function in the wrapper module which gets called at the toplevel (suggested by shunting314). The purpose of this function is to load the constants (torchbind objects, constant tensors). Previously, this code was only in the benchmark_compiled_module function, to prevent the constants from being loaded on every `call`. However, that function is not called in the jit inductor path, causing the torchbind constants to not be loaded. So instead, I added the `load_constants` function which gets called in the toplevel module once.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1234027981 (without benchmark_compiled_module) P1234029088 (with benchmark_compiled_module)

I also added a `load_constants` function in the wrapper module which gets called at the toplevel (suggested by shunting314). The purpose of this function is to load the constants (torchbind objects, constant tensors). Previously, this code was only in the benchmark_compiled_module function, to prevent the constants from being loaded on every `call`. However, that function is not called in the jit inductor path, causing the torchbind constants to not be loaded. So instead, I added the `load_constants` function which gets called in the toplevel module once.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: P1234027981 (without benchmark_compiled_module) P1234029088 (with benchmark_compiled_module)

I also added a `load_constants` function in the wrapper module which gets called at the toplevel (suggested by shunting314). The purpose of this function is to load the constants (torchbind objects, constant tensors). Previously, this code was only in the benchmark_compiled_module function, to prevent the constants from being loaded on every `call`. However, that function is not called in the jit inductor path, causing the torchbind constants to not be loaded. So instead, I added the `load_constants` function which gets called in the toplevel module once.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on test/inductor/test_torchbind.py:
```diff
@@ -0,0 +1,63 @@
+# Owner(s): ["module: functorch"]
```
`test/dynamo_expected_failures/TestTorchbind.test_torchbind_inductor`

New failure here - expected ?

angelayi replied:
Yeah, this is expected since dynamo can't trace the torchbind objects yet.

eellison reviewed (commented):
Inductor sets tensor constants directly onto the python module: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codecache.py#L2576

We should be doing something similar with torchbind constants. The pickling is nice to have so to generate benchmark module but it should not be executed outside of that.

angelayi commented:
> Inductor sets tensor constants directly onto the python module

ohh thanks for the pointer! didnt know about this!

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: [P1245776497](https://www.internalfb.com/phabricator/paste/view/P1245776497)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison reviewed (approved):
Looks good ! it's a little confusing that we have the torchbind constant as `tensor_constant0` in the generated code. would be clearer as `torchbind_constant0` 

Also the equality thing is a unintuitive, would you mind filing an issue to track that ?

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: [P1245776497](https://www.internalfb.com/phabricator/paste/view/P1245776497)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi committed:
Update on "[torchbind] Add inductor support"

Example inductor generated python code: [P1245776497](https://www.internalfb.com/phabricator/paste/view/P1245776497)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

angelayi commented:
@pytorchbot merge
